### PR TITLE
Fix duplicate removal when merging completion suggestions

### LINKED ISSUE #35836: skip_duplicates ignores weight
**Elasticsearch version** (`bin/elasticsearch --version`): 6.5.1, Build: default/tar/8c58350/2018-11-16T02:22:42.182257Z, JVM: 11.0.1

**Plugins installed**: []

**JVM version** (`java -version`):
openjdk version "11.0.1" 2018-10-16
OpenJDK Runtime Environment 18.9 (build 11.0.1+13)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.1+13, mixed mode)

**OS version** (`uname -a` if on a Unix-like system):
Linux 1f3aa3980dfa 4.9.125-linuxkit #1 SMP Fri Sep 7 08:20:28 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:

Running a suggest query with `skip_duplicates` set to `true` ignores `weight` of documents.

**Steps to reproduce**:

Create mapping:

```
PUT /listing_address
{
  "mappings": {
    "listing": {
      "properties": {
        "suggest": {
          "type": "completion",
          "analyzer": "standard",
          "contexts": [
          	{
          		"name": "status_and_rental",
          		"type": "category"
          	}
          ]
        }
      }
    }
  }
}
```

Index some documents:

```
PUT /listing_address/listing/123123123
{
"suggest": {
    "contexts": {
        "status_and_rental": "U"
    },
    "input": [
        "502-30 Church St, Toronto, ON, M5E1S7",
        "30 Church St Toronto ON M5E1S7"
    ],
    "weight": 662
}
}

PUT /listing_address/listing/123123124
{
"suggest": {
    "contexts": {
        "status_and_rental": "UR"
    },
    "input": [
        "502-30 Church St, Toronto, ON, M5E1S7",
        "30 Church St Toronto ON M5E1S7"
    ],
    "weight": 710
}
}

PUT /listing_address/listing/123123125
{
"suggest": {
    "contexts": {
        "status_and_rental": "U"
    },
    "input": [
        "502-30 Church St, Toronto, ON, M5E1S7",
        "30 Church St Toronto ON M5E1S7"
    ],
    "weight": 1228
}
}
```

"status_and_rental" varies from document to document with possible values of "U" or "UR". Weight varies between 662 and 1228.

Here's my search query:

```
POST /listing_address/_search?pretty
{
  "suggest": {
	 "address-suggest": {
    	"prefix": "502-30 church",
		"completion": {
			"field": "suggest",
			"size": 99,
			"skip_duplicates": true,
			"contexts": {
				"status_and_rental": ["UR", "U"]
			}
		}
      }
  }
}
```

The query returns one document with the weight of 662. But there are other documents with higher weight. I would expect the document with the highest weight to be returned.

### PR DESCRIPTION
**Pull request by jimczi**
The completion suggester ignores the original weight of the suggestion when duplicates are removed. This change fixes this bug and keeps the best weighted suggestion among the duplicates. It also removes the custom implementation of the top docs suggest collector now that https://issues.apache.org/jira/browse/LUCENE-8529 is committed in Lucene.

Closes #35836

### CODE REVIEW

**Review by javanna — APPROVED** [review]
I left a couple of minor nits and one suggestion. LGTM though.

File: server/src/test/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionTests.java [hunk 1]
```diff
@@ -61,4 +63,41 @@ public void testToReduce() throws Exception {
...
+                String surfaceForm = randomFrom(surfaceForms);
+                CompletionSuggestion.Entry.Option newOption =
+                    new CompletionSuggestion.Entry.Option(j, new Text(surfaceForm), maxScore - j, Collections.emptyMap());
+                entry.addOption(newOption);
+                options.add(newOption);
+            }
+            shardSuggestions.add(suggestion);
+        }
+        List<CompletionSuggestion.Entry.Option> expected = options.stream()
+            .sorted((o1, o2) -> COMPARATOR.compare(o1, o2))
```
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/test/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionTests.java]: ++

File: server/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestion.java [hunk 2]
```diff
@@ -180,37 +178,43 @@ public static CompletionSuggestion reduceTo(List<Suggest.Suggestion<Entry>> toRe
...
-                        }
-                        if (option == priorityQueue.insertWithOverflow(option)) {
-                            // if the current option has overflown from pq,
-                            // we can assume all of the successive options
-                            // from this shard result will be overflown as well
-                            break;
-                        }
+                    Iterator<Entry.Option> it = ((CompletionSuggestion) suggestion).getOptions().iterator();
+                    if (it.hasNext()) {
+                        pq.add(new SuggestionRef(it));
```
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestion.java]: I agree that `updateTop` is better so I [updated](https://github.com/elastic/elasticsearch/pull/36996/commits/9b1dd2ed1733c8f41d46b51773a7add500b4ece2) the pr with your suggestion

File: server/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestion.java [hunk 3]
```diff
@@ -180,37 +178,43 @@ public static CompletionSuggestion reduceTo(List<Suggest.Suggestion<Entry>> toRe
...
+                        ref.next();
+                        pq.add(ref);
+                    }
+                    if (leader.skipDuplicates &&
+                            seenSurfaceForms.add(current.getText().toString()) == false) {
+                        continue;
+                    }
+                    options.add(current);
+                    if (options.size() >= size) {
+                        break;
```
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestion.java]: sure, this makes sense. I pushed https://github.com/elastic/elasticsearch/pull/36996/commits/9b1dd2ed1733c8f41d46b51773a7add500b4ece2

**Pull request merged**